hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -287,19 +285,25 @@ object InferPType {
         InferPType(nd, env)
         InferPType(shape, env)
 
-        PNDArray(coerce[PNDArray](nd.pType2).elementType, shape.pType2.asInstanceOf[TTuple].size, nd.pType2.required)
+        PNDArray(coerce[PNDArray](nd.pType2).elementType, shape.pType2.asInstanceOf[PTuple].size, nd.pType2.required)
       }
       case NDArrayMap(nd, name, body) => {
         InferPType(nd, env)
-        InferPType(body, env.bind(name, nd.pType2))
+        val ndPType = nd.pType2.asInstanceOf[PNDArray]
+        InferPType(body, env.bind(name -> ndPType.elementType))
 
-        PNDArray(body.pType2, coerce[PNDArray](nd.pType2).nDims, nd.pType2.required)
+        PNDArray(body.pType2, ndPType.nDims, body.pType2.required)
```

Comment:
This I now think is wrong, but it follows our ArrayMap implementation. If body is not required, and ends up being empty, but the left hand operand is required, is there an array constructed/allocated?

We have a somewhat, or at least seemingly, conflicting message with ArrayFlatMap, line 242: 
"Whether an array must return depends on a, but element requiredeness depends on body (null a elements elided)"